man/man1/pmcheck.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2024 Ken McDonell.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMCHECK 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmcheck\f1 \- manage the configuration of optional PCP components
.SH SYNOPSIS
\f3$PCP_BINADM_DIR/pmcheck\f1
[\f3\-a\f1] [\f3\-d\f1] [\f3\-l\f1] [\f3\-n\f1] [\f3\-s\f1]
[\f3\-v\f1] [\f3\-x\f1] [component ...]
.br
\f3$PCP_BINADM_DIR/pmcheck\f1
[\f3\-a\f1] [\f3\-c\f1 script]] [\f3\-d\f1] [\f3\-l\f1] [\f3\-n\f1] [\f3\-s\f1]
[\f3\-v\f1] [\f3\-x\f1]
.br
\f3pcp-check\f1 ...
.SH DESCRIPTION
The Performance Co-Pilot (PCP) includes a number of core components
that are likely to be pre-configured and operational, but it also
includes a larger number of components that are not active by default.
In the context of
.B pmcheck
these components cover services that are typically started as part
of the system boot procedure, e.g.
.BR pmcd (1),
.BR pmlogger (1),
.BR pmproxy (1),
etc.\& or optional
Performance Metric Domain Agents (PMDAs) that augment the available
performance metrics exported by
.BR pmcd (1).
.PP
.B pmcheck
allows the interrogation of the state of components with
the \fB\-s\fR, \fB\-\-state\fR option.
Components can be
activated with the
\fB\-a\fR, \fB\-\-activate\fR option,
or deactivated with the
\fB\-d\fR, \fB\-\-deactivate\fR option.
These three options are clearly mutually exclusive for any single
execution of
.BR pmcheck ,
and in the absence of any of them
.B \-s
is assumed to be the default.
.PP
The set of components to be interrogated or configured are specified
by the
.I component
name(s) from the command line, else all known components in the absence
of any
.I component
arguments.
When one or more
.I component
arguments are specified, these may optionally be
.BR sh (1)
(glob) patterns that are matched against the names of the known
components.
.PP
The \fB\-l\fR, \fB\-\-list\fR option
reports on known components.
With one or more
.I components
each is reported.
Without any
.I component
arguments the
.I \-l
option lists all known components,
which is simply all the component scripts (see the
.B "COMPONENT SCRIPTS"
section below)
stored in the
.I $PCP_SHARE_DIR/lib/pmcheck
directory.
If the
.B \-v
option is also used, a short description of each requested component is also
displayed.
.PP
Additional components can be integrated into the
.B pmcheck
framework, and the
\fB\-c\fR, \fB\-\-file\fR option
allows an alternate
.BR sh (1)
.I script
to be used instead of a script associated with a known
.IR component .
.PP
In concert with the
.B \-a
or
.B \-d
options,
the \fB\-n\fR, \fB\-\-show-me\fR option performs a dry run,
showing the
.BR sh (1)
commands that would be needed to perform the reconfiguration,
but no reconfiguration is done.
Commands that need to be run as ``root'' are prefixed with
.B #
and other commands are prefixed with
.BR $ .
.PP
The \fB\-v\fR, \fB\-\-verbose\fR option increases reporting
verbosity.
.PP
The \fB\-x\fR, \fB\-\-trace\fR option may be used specify that
the component scripts are run with
.B \-x
option to
.BR sh (1).
This can assist with debugging the failure of a component script
or during the development of a new component script, probably
also involving the
.B \-c
option to
.BR pmcheck .
.PP
The \fB\-?\fR, \fB\-\-help\fR
option displays a usage message.
.PP
As a convience,
.B pcp-check
is an alias for
.B pmcheck .
.SH COMPONENT SCRIPTS
The
.B pmcheck
framework is extensible, as the set of known components is defined
by the names of the component scripts in the
.I $PCP_SHARE_DIR/lib/pmcheck
directory.
.PP
Each component script needs to abide by the following rules
for correct integration with
.BR pmcheck :
.IP \(bu 2n
.PD 0
The script will be run with
.BR sh (1),
if you need some other scripting language then you must provide a
.BR sh (1)
wrapper.
.IP \(bu
Since each script is responsible for one component, the name
of the component is implied, however the name of the component
is also passed from
.B pmcheck
as the sole argument after the options.
This allows generic component scripts to be implemented as hard
links below
.IR $PCP_SHARE_DIR/lib/pmcheck .
.IP \(bu
When called with a
\fB\-l\fR, \fB\-\-list\fR option the script does nothing,
unless it is also called with a
.B \-v
option in which case
a short description of the component is displayed on standard output.
.IP \(bu
When called with a
\fB\-s\fR, \fB\-\-state\fR option, the script must exit with the an
exit code that indicates the component's state, namely 0
(active), 1 (inactive, but
.I could
be activated by
.BR pmcheck ),
2 (inactive and
.I cannot
be activated by
.BR pmcheck ).
The distinction between exit codes 1 and 2 is based on many factors,
but 1 usually means all of the required software packages (PCP and
other) are installed,
required prerequisite services are running or could be safely started,
and any component configuration can assume default settings or can
be reliably deduced, e.g. a port number for communications.
Conversely, 2 means automatic activation of the component is not
possible and some end-user intervention is required.
.IP \(bu
When called with a \fB\-a\fR, \fB\-\-activate\fR option
the script must activate the component.
This means making any required configuration change so that it
.I is
restarted at the next system reboot or
.BR pmcd (1)
restart, and
starting the component if it is currently inactive,
.IP \(bu
When called with a \fB\-d\fR, \fB\-\-deactivate\fR option
the script must deactivate the component.
This means stopping the component if it is currently active,
and making any required configuration change so that it
.I "is not"
restarted at the next system reboot or
.BR pmcd (1)
restart.
.IP \(bu
When called with a \fB\-n\fR, \fB\-\-show-me\fR option
the script must perform a dry run and echo any required
commands on standard output as per the rules described above
for the
.B \-n
option of
.BR pmcheck .
.IP \(bu
If a \fB\-v\fR, \fB\-\-verbose\fR option is also specified, the
script should provide explanatory text on standard output to help
explain, particularly in cases where the component is not active
or the reconfiguration failed.
.PD
.PP
To assist with component script development, the file
.I $PCP_SHARE_DIR/lib/checkproc.sh
may be sourced (after sourcing
.IR $PCP_DIR/etc/pcp.env ).
.I checkproc.sh
provides the following services:
.IP \(bu 2n
.PD 0
A temporary directory is created and the shell variable
.BR $tmp
is assigned to the full pathname of the directory.
.IP \(bu
The shell variable
.B $status
is used to return the exit status from the script [defaults to 0].
.IP \(bu
The shell variable
.B $component
is assigned to the associated component name.
.IP \(bu
The shell variable
.B $prog
is assigned to the script's basename (so the component name
in the default case).
.IP \(bu
A
.B trap
is installed to cleanup
.B $tmp
and exit with
.BR $status .
.IP \(bu
A
.B _do_args
shell procedure that handles all command argument parsing.
.IP \(bu
Shell variables
.BR $aflag ,
.BR $dflag ,
.BR $lflag
or
.B $sflag
have default values of
.B false
and are set to
.B true
by
.B _do_args
when the
.BR \-a ,
.BR \-d ,
.BR \-l
or
.B \-s
options are parsed.
.IP \(bu
The shell variable
.B $show_me
has the default value of
.B false
and is set to
.B true
by
.B _do_args
when the
.B \-n
option is parsed.
.IP \(bu
The shell variable
.B $verbose
counts number of
.B \-v
options parsed by
.BR _do_args .
.IP \(bu
A
.B _ctl_svc
shell procedure that handles status requests
and all reconfiguration operations for
services that are managed by
.BR systemctl (1)
or
.BR init (1)
or similar infrastructure layers.
.B _ctl_svc
takes 2 arguments, the first is an
.I action
(one of
.BR state ,
.BR start ,
.BR stop ,
.B activate
(enable and start),
or
.B deactivate
(stop and disable),
and the second is the
.I name
of the component;
for
.BR systemctl (1)
managed components,
.I name
should be the basename of the service, e.g.
.B pmcd
for the
.B pmcd.service
.BR systemd (1)
unit, otherwise
.I name
is the name of the
.BR init (1)
or similar ``rc'' script.
.IP \(bu
.B _ctl_svc
will return a value (via
.B $?
when called from a shell script) that matches the exit codes
defined above for each action.
.IP \(bu
A
.B _ctl_pmda
shell procedure that handles status requests
and all reconfiguration operations for a PMDA
that is connected to
.BR pmcd (1).
.B _ctl_pmda
takes at least 2 arguments, the first is an action (one of
.BR state ,
.B activate
or
.BR deactivate ,
and the second argument is the name of the PMDA.
For the
.B install
action, there is a third argument for the name of the
PMDA's executable or DSO and an optional fourth argument
is the name of a file providing the input required for
the PMDA's
.I Install
script (defaults to /dev/null).
.IP \(bu
.B _ctl_pmda
will return a value (via
.B $?
when called from a shell script) that matches the exit codes
defined above for each action.
.PD
.PP
So the following is the simplest possible component script for
a system service.
.PP
.ft CR
.nf
#!/bin/sh
\&. $PCP_DIR/etc/pcp.env || exit 1
\&. $PCP_SHARE_DIR/lib/checkproc.sh
.sp 0.5v
_do_args "$@"
.sp 0.5v
if $lflag
then
    [ $verbose -gt 0 ] && echo "my description"
elif $sflag
then
    _ctl_svc state $component
    status=$?
elif $aflag
then
    _ctl_svc activate $component || status=1
elif $dflag
then
    _ctl_svc deactivate $component || status=1
fi
.fi
.ft
.PP
And the following is the simplest possible component script for
a PMDA.
.PP
.ft CR
.nf
#!/bin/sh
\&. $PCP_DIR/etc/pcp.env || exit 1
\&. $PCP_SHARE_DIR/lib/checkproc.sh
.sp 0.5v
_do_args "$@"
.sp 0.5v
if $lflag
then
    [ $verbose -gt 0 ] && echo "my description"
elif $sflag
then
    _ctl_pmda state $component
    status=$?
elif $aflag
then
    _ctl_pmda activate $component || status=1
elif $dflag
then
    _ctl_pmda deactivate $component || status=1
fi
.fi
.ft
.SH DIAGNOSTICS
For
.B pmcheck
(not the component scripts) the exit status is typically 0, but values of 9 or 99 indicate serious
internal botches.
.PP
When the
.B \-a
or
.B \-d
options are used, the success or failure of the reconfiguration is
reported on standard output.
.PD
.SH CAVEAT
Some components have prerequisites that demand additional non-PCP services be
installed, configured and operational, e.g. a database server or a key-value
server.
In these cases, probing the status of the non-PCP service may be subject to
some access controls and if
.B pmcheck
is run as a non-root user, then the output from
.B \-s
(the default)
may be incomplete or misleading; in these circumstances, running
.B pmcheck
as root should provide a more accurate report.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR init (1),
.BR PCPIntro (1),
.BR pmcd (1),
.BR pmlogger (1),
.BR pmproxy (1),
.BR sh (1),
.BR systemctl (1),
.BR pcp.conf (5)
and
.BR pcp.env (5).

.\" control lines for scripts/man-spell
.\" +ok+ _ctl_pmda _ctl_svc _do_args aflag checkproc dflag
.\" +ok+ gt {from test -gt}  lflag pre prog rc {from ``rc'' script}
.\" +ok+ sflag show_me tmp


CHANGELOG
pcp-6.2.2 (15 May 2024)
    - pmdalibvirt: add support for metric labels
    - pmdalibvirt: new balloon, vCPU and domain info metrics
    - pmdalibvirt: remove support for the ancient RHEL 5 era "oldapi"
    - pmdalinux: implement mem.vmmemctl VM balloon metrics
    - pmdauwsgi: improve handling of transient uWSGI servers
    - pmdauwsgi: short-circuit handling of pmcheck failure
    - pmdauwsgi: fix semantics of several metrics
    - pmcd: rate-limit connection failure logging in case of bursts
    - pmcheck: rework checkproc.sh when used in non-systemd environment
    - pmrep conf: minor tweaks to sync with sysstat 12.7.5
    - pmrep conf: fix few pidstat metric formulas
    - pmlogdump: fix -V/--version option handling
    - pmlogger_daily_report.sh: fix pmrep error filtering
    - pcp2openmetrics: minor bug fixes, added archive timestamps
    - indomcachectl: new tool to initialize/dump indom cache files
    - libpcp_pmda: add PMDA_CACHE_WRITE indom cache operation
    - libpcp_web: ensure openmetrics scrape sees all labels
    - libpcp: derived metrics: add novalue() and defined(x) rework
    - libpcp: derived metrics lazy binding for defined(x) ternary
    - libpcp: add __pmNotifyThrottle and __pmResetNotifyThrottle
    - python API: pass help text result as string not bytes
    - python API: fix return value from pmLoadDerivedConfig
    - tmpfiles.d: extend the tmpfiles.d directory set to logdirs
    - bash,zsh: ensure completion for both pmlogdump and pmdumplog
    - build: add missing python rpms to pcp-testsuite deps
    - build: use Y2038-safe time_t interfaces on 32-bit platforms
    - selinux: allow pmie/pmlogger services to use systemd status
    - selinux: ignore pmie/pmlogger getattr on nsfs files
    - docs: add pmrep(1) hint about CSV delimiter with metricsets
    - man pages: pmnewcontext.3: reword PM_CONTEXT_LOCAL description
    - man pages: man-spell updates

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging

src/pmcheck/pmcheck.sh
#!/bin/sh
#
# Check/report-status/enable/disable/ PCP components
#

# common initialization
#
. $PCP_DIR/etc/pcp.env || exit 1

status=0
prog=`basename $0`
tmp=`mktemp -d "$PCP_TMPFILE_DIR/pmcheck.XXXXXXXXX"`
if [ ! -d "$tmp" ]
then
    echo >&2 "$prog: Error: cannot create temp directory: $PCP_TMPFILE_DIR/$prog.XXXXXXXXX"
    exit 9
fi
trap "rm -rf $tmp; exit \$status" 0 1 2 3 15

# common command line processing ... use pmgetopt(1) for portability
#
aflag=false
cfile=''
dflag=false
lflag=false
show_me=false
sflag=false
verbose=0
xflag=false
action=''

cat <<'End-of-File' >$tmp/_usage
# getopts: ac:dlnsvx?
# usage: [options] [component ...]

options:
  -a, --activate    activate component(s)
  -c=SCRIPT, --file=SCRIPT use this executable SCRIPT file instead of a standard component
  -d, --deactivate  deactivate component(s)
  -l, --list        list components
  -n, --show-me     dry run
  -s, --state       report state of component(s)
  -v, --verbose     increase verbosity
  -x, --trace       run component script with sh -x
  -?, --help        show this usage message
# end

component is one or more manageable components and may include shell
glob metacharacters, e.g. pmda.* or pm[cl]*
End-of-File

__args=`pmgetopt --progname=$prog --config=$tmp/_usage -- "$@"`
if [ $? -ne 0 ]
then
    pmgetopt --progname=$prog --config=$tmp/_usage --usage
    status=9
    exit
fi

eval set -- "$__args"
unset __args

while [ $# -gt 0 ]
do
    case "$1"
    in
	-a)	# activate
		aflag=true
		action="$action -a"
		;;
	-c)	# alternate component script
		if [ ! -f "$2" ]
		then
		    echo >&2 "$prog: Error: \"$2\" does not exist"
		    status=9
		    exit
		fi
		if [ ! -x "$2" ]
		then
		    echo >&2 "$prog: Error: \"$2\" must be an executable script"
		    status=9
		    exit
		fi
		cfile="$2"
		shift
		;;
	-d)	# deactivate
		dflag=true
		action="$action -d"
		;;
	-l)	# list
		lflag=true
		action="$action -l"
		;;
	-n)	# dry run
		show_me=true
		action="$action -n"
		;;
	-s)	# state
		sflag=true
		action="$action -s"
		;;
	-v)	# verbose
		verbose=`expr $verbose + 1`
		action="$action -v"
		;;
	-x)	# sh -x tracing
		xflag=true
		;;
	--)	# end of opts
		shift
		break
		;;
	-\?)	# help or error
		pmgetopt --progname=$prog --config=$tmp/_usage --usage
		exit
		;;
	-\*)	# botch
		echo >&2 "$prog: Error: unrecognized option ($1)"
		status=9
		exit
    esac
    shift
done

# semantic checks on command line arguments
#
if [ -n "$cfile" -a $# -gt 0 ]
then
    echo >&2 "$prog: Error: option -c and component arguments are mutually exclusive"
    status=9
    exit
fi
if [ "$aflag" = false -a "$dflag" = false -a "$show_me" = true ]
then
    echo >&2 "$prog: Warning: neither -a nor -d specified, so -n ignored"
    show_me=false
fi

nop=0
$aflag && nop=`expr $nop + 1`
$dflag && nop=`expr $nop + 1`
$lflag && nop=`expr $nop + 1`
$sflag && nop=`expr $nop + 1`
if [ "$nop" -eq 0 ]
then
    # default to -s
    #
    sflag=true
    action="$action -s"
fi
if [ "$nop" -gt 1 ]
then
    echo >&2 "$prog: Error: at most one of -a, -d, -l or -s may be specified"
    status=9
    exit
fi

if [ $# -eq 0 ]
then
    # no components on the command line, do 'em all
    #
    rm -f $tmp/tmp
    set -- `find $PCP_SHARE_DIR/lib/pmcheck -type f \
	    | sed \
		-e "s@$PCP_SHARE_DIR/lib/pmcheck/@@" \
	    | LC_COLLATE=POSIX sort`
    if [ -s $tmp/tmp ]
    then
	echo >&2 "$prog: Warning: ignoring component files from package installation: `cat $tmp/tmp`"
	rm -f $tmp/tmp
    fi
fi

# reporting for -a, -d and -s
#
_report()
{
    if $show_me
    then
	cat $tmp/out
	return
    fi

    if $sflag
    then
	case "$2"
	in
	    0)	msg="active"
	    	;;
	    1)	msg="could be activated"
	    	;;
	    2)	msg="cannot be activated"
	    	;;
	    *)	msg="botch, rc=$2"
	    	;;
	esac
    else
	case "$2"
	in
	    0)	msg="success"
	    	;;
	    1)	msg="failure"
	    	;;
	    *)	msg="botch, rc=$2"
	    	;;
	esac
    fi
    printf "%-15s %-20s" "$1" "$msg"
    if [ -s $tmp/out ]
    then
	first=true
	cat $tmp/out \
	| while read line
	do
	    if $first
	    then
		printf " %s\n" "$line"
		first=false
	    else
		printf "%-15s %-20s %s\n" "" "" "$line"
	    fi
	done
    else
	printf "\n"
    fi
}

# reporting for -l
#
_report_list()
{
    printf "%-15s" "$1"
    if [ -s $tmp/out ]
    then
	first=true
	cat $tmp/out \
	| while read line
	do
	    if $first
	    then
		printf " %s\n" "$line"
		first=false
	    else
		printf "%-15s %s\n" "" "$line"
	    fi
	done
    else
	printf "\n"
    fi
}

if [ "$cfile" != "" ]
then
    if $xflag
    then
	sh -x "$cfile" $action >$tmp/out
    else
	"$cfile" $action >$tmp/out
    fi
    rc=$?
    _report "$cfile" $rc
    exit
fi

if $lflag
then
    # list components
    #
    for component
    do
	script="$PCP_SHARE_DIR/lib"/pmcheck/$component
	if [ ! -x "$script" ]
	then
	    echo "$prog: Error: $script not found or not executable"
	else
	    rm -f $tmp/out
	    if [ $verbose -gt 0 ]
	    then
		if $xflag
		then
		    sh -x "$script" $action $component >$tmp/out
		else
		    "$script" $action $component >$tmp/out
		fi
	    fi
	    _report_list "$component"
	fi
    done
    exit
fi

# actions with installed components
#
for pattern in "$@"
do
    for script in "$PCP_SHARE_DIR/lib"/pmcheck/$pattern
    do
	component=`basename "$script"`
	if [ ! -x "$script" ]
	then
	    echo "$prog: Error: $script not found or not executable"
	else
	    if $xflag
	    then
		sh -x "$script" $action $component >$tmp/out
	    else
		"$script" $action $component >$tmp/out
	    fi
	    rc=$?
	    _report "$component" $rc
	fi
    done
done
